### DISCUSSION TITLE
In the Popup Menu in Sirius, how to display only the first-level menu

### DISCUSSION THREAD

**Discussion by hoajie** [eclipse-sirius/sirius-desktop#453] [Q&A] [answered]
The Popup Menu (right-click menu) in Sirius can provide users with operation options. To display only the first-level menu (not the sub-menu)
When I use the Popup Menu (right-click menu) in Sirius, I want users to use only the first-level menu, such as the example of 'Rename' in the picture - users can start editing by clicking 'Rename'. But now Sirius must have a sub-menu, that is, users need to click Rename->Rename to edit the element name - that is, users need to click twice. This does not meet my work requirements. Can I hide the sub-menu so that users can edit the element name with just one click?
![Snipaste_2024-09-06_15-50-03](https://github.com/user-attachments/assets/4c0bf983-e74f-4f78-87f0-8eed942855cc)

**Comment by lredor** [comment, marked as answer]
As you can see in the data used for tests, it is possible to have an action without "sub-menu". It will be at the end of the popup as in this screenshot:
![image](https://github.com/user-attachments/assets/847b11f6-9b4d-45b9-9a0b-b4a825980843)
The properties of the Popup Menu is like that (without id in the field "location URI") :
![image](https://github.com/user-attachments/assets/f958bc93-5acf-4ede-84ae-6f93a9120da6)

The data of tests is available [here](https://github.com/eclipse-sirius/sirius-desktop/tree/master/plugins/org.eclipse.sirius.tests.swtbot/data/unit/popupMenu).

It seems also possible to add an "Operation Action" directly at the root of the tools section.

**Reply by t-chiang** [reply]
Hello, I'm trying to work based on the examples in the tests. I'm trying to create context menu options similar to what is shown however I'm unsure what is supposed to go into the Location URI. I'm trying to use the preconditions to activate the menu but have been unsuccessful so far. Do the examples still use extension points to create the context menu options or are the only specified using the odesign file?

![image](https://github.com/user-attachments/assets/ecd9e3f7-218a-4fc8-a213-112b315c3131)

![image](https://github.com/user-attachments/assets/bd0f8afa-657f-4844-9628-6fcb9f7e6bde)

I want to create a context menu option that uses the tools and querying present in Sirius and don't want to have to use Java extension points to define the functions.

**Comment by lredor** [comment]
I have created a sample project from the test data. I reduced the number of menus and added preconditions. These actions only use the VSM (no extension point). In the sample, the actions are available on EPackage on not on EClass, EAttribute and EReference.
[SamplePopupMenuProject.zip](https://github.com/user-attachments/files/17323832/SamplePopupMenuProject.zip)
![2024-10-10_09h09_54](https://github.com/user-attachments/assets/174255ec-a661-4dad-81d0-ad63889ead75)
![2024-10-10_09h10_04](https://github.com/user-attachments/assets/ffa54589-8299-4e42-9a60-f983c9e75653)

**Reply by lredor** [reply]
This project replaces the previous one (that contains a "version" problem).
[SamplePopupMenuProject.zip](https://github.com/user-attachments/files/17327563/SamplePopupMenuProject.zip)

**Comment by lredor** [comment]
[hoajie](https://github.com/hoajie) Fill free to reopen if you have additional question. For information, your question seems similar to #476.

**Reply by hoajie** [reply]
Hi lredor,

Thank you for your response, and I apologize for the delayed reply. I appreciate the answer and the reference to issue #476, which I'll review to see if it resolves my question. If I have any additional questions, I'll be sure to follow up.

Thanks again for your help!

Best regards,  
haojie
